hack/oci-pin/internal/sums/sums.go

```diff
@@ -0,0 +1,225 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+// sums creates and combines sha256sums files.
+package sums
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"path"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+// Create creates a sha256sums file from a list of digests.
+// The digests going in are expected to be in the oci format "sha256:hex".
+func Create(refs []PinnedImageReference, out io.Writer) error {
+	normalizedRefs, err := normalizeRefs(refs)
+	if err != nil {
+		return err
+	}
+	for _, ref := range normalizedRefs {
+		if _, err := fmt.Fprintf(out, "%s  %s\n", ref.Sum(), ref.Reference()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Merge merges multiple sha256sums files into one.
+func Merge(refs [][]PinnedImageReference, out io.Writer) error {
+	mergedRefs := make([]PinnedImageReference, 0)
+	for _, refs := range refs {
+		mergedRefs = append(mergedRefs, refs...)
+	}
+	return Create(mergedRefs, out)
+}
+
+// Parse parses a sha256sums file.
+func Parse(in io.Reader) ([]PinnedImageReference, error) {
+	scanner := newLineScanner(in)
+	return parseLines(scanner)
+}
+
+// PinnedImageReference contains the components of a pinned image reference.
+type PinnedImageReference struct {
+	Registry string
+	Prefix   string
+	Name     string
+	Tag      string
+	Digest   string
+}
+
+// Reference returns the string representation of the reference (without the digest).
+func (r PinnedImageReference) Reference() string {
+	var base string
+	if r.Prefix == "" {
+		base = path.Join(r.Registry, r.Name)
+	} else {
+		base = path.Join(r.Registry, r.Prefix, r.Name)
+	}
+	var tag string
+	if r.Tag != "" {
+		tag = ":" + r.Tag
+	}
+	return base + tag
+}
+
+// Sum returns the string representation of the digest.
+// The digest is expected to be in the oci format "sha256:hex".
+// The resulting Sum is only the hex part.
+func (r PinnedImageReference) Sum() string {
+	return r.Digest[len("sha256:"):]
+}
+
+// normalizeRefs normalizes the image references.
+// It sorts the references and removes duplicates.
+// If conflicting digests are found, an error is returned.
+func normalizeRefs(refs []PinnedImageReference) ([]PinnedImageReference, error) {
```

Comment:
Wouldn't `coalesceRefs` be more precise, since the may reduce the number of entries. The formatting of the entries itself is not touched/normalized.